replication.go

```diff
@@ -586,12 +595,33 @@ func (r *Raft) handleStaleTerm(s *followerReplication) {
 // updateLastAppended is used to update follower replication state after a
 // successful AppendEntries RPC.
 // TODO: This isn't used during InstallSnapshot, but the code there is similar.
-func updateLastAppended(s *followerReplication, req *AppendEntriesRequest) {
+func (r *Raft) updateLastAppended(s *followerReplication, req *AppendEntriesRequest) {
 	// Mark any inflight logs as committed
 	if logs := req.Entries; len(logs) > 0 {
 		last := logs[len(logs)-1]
 		atomic.StoreUint64(&s.nextIndex, last.Index+1)
 		s.commitment.match(s.peer.ID, last.Index)
+
+		// If peer is "caught up" ensure it's not longer considered "recovering". We
+		// somewhat arbitrarily define caught up to mean that it's at least half way
+		// into the current TrailingLogs buffer so that if log compaction happens
+		// right after this, it will still likely be able to continue catching up.
+		lastLog, _ := r.getLastLog()
+		delta := r.conf.TrailingLogs / 2
+		caughtUp := last.Index > (lastLog - delta)
+
+		// Take lock and see if we need to modify recoveringPeer state.
+		r.leaderState.recoveringPeersLock.Lock()
+		if r.leaderState.recoveringPeers != nil {
+			if caughtUp {
+				delete(r.leaderState.recoveringPeers, s.peer.ID)
```

Comment:
It seems that items in `recoveringPeers` should also be able to expire some other way. If a recovering peer drops off before catching up, and never comes back, the logs will never be truncated beyond that peer's last known index.

Reply:
Yeah great point. I wonder how to resolve this. If it's only time based then it comes back to "how long is it reasonable to expect a snapshot to take" which is kind of the same problem we are trying to solve here. i.e. any timeout even if it's configurable is just another upper bound on how big snapshots can get relative to restore speed before you get in the same unrecoverable state.

That said, it might still make sense to cap it _somewhere_ and at least it keeps the common case logs smaller but allows recovery to increase storage only if needed up to some much higher limit? But what is that limit? Maybe an hour? That's a lot more than the recovery time of the one case we've seen in prod now but it's already the case that they are operating outside of our design envelope so it's hard to come up with a non-arbitrary limit on how far outside we want to support? And that's only Consul's use case. Other DBs might use this raft implementation for cases where much larger datasets are expected.

For example I used to have to recover 200GB+ MongoDB replica sets that had exactly this problem since they have a fixed-size OpLog. In some cases they took several hours to recover from snapshot. I guess if it's configurable then people can configure it up for that kind of use case (although it doesn't sound like a great idea to me!).

Reply:
Couldn't we just remove the entry from recoveringPeers when that peer is removed from the Raft Configuration?

Reply:
Maybe we could kick off some arbitrary timer when the peer is removed from the Raft config. That way we are less dependent on the time it takes to transfer snapshots, and it would account for the situation where the peer restarts. 

When the timer elapses, if the peer is still there we update the index, if it's not we delete the entry from recoveringPeers.

Reply:
Removing from raft config is a good call.

I guess there is still the case where a follower gets stuck doing something else though and doesn't cleanly crash or get removed from the quorum.

After thinking about this some, I decided we can probably just be more conservative than this PR and instead:

 - if `installSnapshot` fails then remove the peer from recovering
 - if it succeeds, set a timer to remove it in X minutes
 - don't bother waiting for TrailingLogs/2 appends - by the time install snapshot has succeeded the follower has not just downloaded but completely restored so we should pretty much immediately attempt appendEntries if we can so having a buffer of a few minutes should be plenty in all but the most pathalogical hanging case and those cases are the ones we want to protect against here I think.

What do you both think?

Reply:
I think the timer approach should work. Should we expose X as a configuration?

Reply:
Maybe, since X could not be enough depending on the workload.

Reply:
On the other hand, if someone is repeatedly running into X not being high enough, then TrailingLogs should probably be bumped up instead.